File: Ancono/src/cn/timelives/java/math/abstractAlgebra/calculator/RingCalculator.java
/**
 * 2018-02-28
 */
package cn.timelives.java.math.abstractAlgebra.calculator;

import cn.timelives.java.utilities.ModelPatterns;

/**
 * The ring calculator defines some basic operations in a group. 
 * @author liyicheng
 * 2018-02-28 18:28
 *
 */
public interface RingCalculator<T> extends GroupCalculator<T> {
	/**
	 * Gets the zero element in this ring, which is the identity element of the addition group.
	 * @return {@code 0}
	 */
	T getZero();
	
	/**
	 * Returns the result of {@code x+y}.
	 * @param x
	 * @param y
	 * @return {@code x+y}
	 */
	T add(T x, T y);
	/**
	 * Returns the negate of this number.
	 * 
	 * @param x
	 *            
	 * @return {@code -x}
	 */
	T negate(T x);
	
	/**
	 * Returns the result of {@code x-y}, which is equal to {@code x+negate(y)}.
	 * @param x
	 * @param y
	 * @return {@code x-y}
	 */
	T subtract(T x,T y);
	
	/**
	 * Returns the result of {@code x*y}. This operation may be not commutative.
	 * @param x
	 * @param y
	 * @return {@code x*y}
	 */
	T multiply(T x,T y);
	
	/**
	 * Return {@code x ^ n} as defined in the multiplicative semigroup.
	 * 
	 * @param x
	 *            
	 * @param n a positive number
	 * @return {@code x ^ n}
	 */
	default T pow(T x, long n) {
		return ModelPatterns.binaryReduce(n, x, this::multiply);
	}
	
	/**
	 * Return the result of {@code n * p}, which is equal to applying addition to 
	 * {@code x} for {@code n} times.
	 * @param x
	 * @param n a long 
	 * @return
	 */
	default T multiplyLong(T x,long n) {
		return gpow(x, n);
	}
	
	/**
	 * @deprecated use {@link #getZero()} instead for more clarity.
	 */
	@Deprecated
	@Override
	default T getIdentity() {
		return getZero();
	}
	
	/**
	 * @deprecated use {@link #add(Object, Object)} instead for more clarity.
	 */
	@Deprecated
	@Override
	default T apply(T x, T y) {
		return add(x,y);
	}
	
	/**
	 * @deprecated use {@link #negate(Object)} instead for more clarity.
	 */
	@Deprecated
	@Override
	default T inverse(T x) {
		return negate(x);
	}
	
	/**
	 * @deprecated use {@link #multiplyLong(Object, long)} instead for more clarity.
	 */
	@Deprecated
	@Override
	default T gpow(T x, long n) {
		return GroupCalculator.super.gpow(x, n);
	}
}

File: src/cn/timelives/java/math/algebra/abstractAlgebra/calculator/RingCalculator.kt
/**
 * 2018-02-28
 */
package cn.timelives.java.math.algebra.abstractAlgebra.calculator

import cn.timelives.java.utilities.ModelPatterns

/**
 * The ring calculator defines some basic operations in a group.
 * @author liyicheng
 * 2018-02-28 18:28
 */
interface RingCalculator<T : Any> : GroupCalculator<T> {
    /**
     * Returns `true` because a ring is always an abelian group by its
     * addition.
     */
    override val isCommutative: Boolean
        get() = true

    /**
     * Determines whether the multiplication is commutative. It is false by default.
     */
    val multiplyIsCommutative : Boolean
        get() = false

    /**
     * Gets the zero element in this ring, which is the identity element of the addition group.
     * @return `0`
     */
    val zero: T

    /**
     * Returns the result of `x+y`.
     * @param x
     * @param y
     * @return `x+y`
     */
    fun add(x: T, y: T): T

    /**
     * Returns the negate of this number.
     *
     * @param x
     *
     * @return `-x`
     */
    fun negate(x: T): T

    /**
     * Returns the result of `x-y`, which is equal to `x+negate(y)`.
     * @param x
     * @param y
     * @return `x-y`
     */
    fun subtract(x: T, y: T): T

    /**
     * Returns the result of `x*y`. This operation may be not commutative.
     * @param x
     * @param y
     * @return `x*y`
     */
    fun multiply(x: T, y: T): T

    /**
     * Return `x ^ n` as defined in the multiplicative semigroup.
     *
     * @param x
     *
     * @param n a positive number
     * @return `x ^ n`
     */
    fun pow(x: T, n: Long): T {
        return ModelPatterns.binaryProduce(n, x) { a, b -> this.multiply(a, b) }
    }

    /**
     * Return the result of `n * p`, which is equal to applying addition to
     * `x` for `n` times. This method can be implemented for better performance.
     * @param x
     * @param n a long
     * @return
     */
    fun multiplyLong(x: T, n: Long): T {
        return super.gpow(x, n)
    }


    @get:Deprecated("use {@link #zero} instead for more clarity.", ReplaceWith("zero"))
    override val identity: T
        get() = zero


    @Deprecated("use {@link #add(Object, Object)} instead for more clarity.", ReplaceWith("add(x, y)"))
    override fun apply(x: T, y: T): T {
        return add(x, y)
    }


    @Deprecated("use {@link #negate(Object)} instead for more clarity.", ReplaceWith("negate(x)"))
    override fun inverse(x: T): T {
        return negate(x)
    }

    @Deprecated("use {@link #multiplyLong(Object, long)} instead for more clarity.", ReplaceWith("super@GroupCalculator.gpow(x, n)"))
    override fun gpow(x: T, n: Long): T {
        return super.gpow(x, n)
    }
}
